File: Java/balanced-binary-tree.java
//  Balanced Binary Tree

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// Time O(N), Space O(1)
class Solution {
    public boolean isBalanced(TreeNode root) {
        return getHeight(root) >= 0;
    }
    
    private int getHeight(TreeNode root) {
        if (root == null)
            return 0;
        int l_height = getHeight(root.left);
        int r_height = getHeight(root.right);
        if (l_height==-1 || r_height==-1 || Math.abs(l_height-r_height)>1)
            return -1;
        return (l_height > r_height) ? (l_height + 1) : (r_height + 1);
    }
}

File: leetcode/balanced-binary-tree.cc
//  Balanced Binary Tree

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if (!root)  return true;
        return get_height(root) >= 0;
    }
    
    int get_height(TreeNode *root) {
        if (!root)  return 0;
        int l = get_height(root->left);
        int r = get_height(root->right);
        if (l==-1 || r==-1 || abs(l - r) > 1)
            return -1;
        else
            return max(l,r) + 1;
    }
};